src/System.Net.Security/src/System/Net/Security/SniHelper.cs

```diff
@@ -0,0 +1,377 @@
+                return null;
+            }
+
+            ExtensionType extensionType = (ExtensionType)BinaryPrimitives.ReadUInt16BigEndian(extension);
+            ReadOnlySpan<byte> extensionData = extension.Slice(ExtensionDataOffset);
+
+            if (extensionType == ExtensionType.ServerName)
+            {
+                return GetSniFromServerNameList(extensionData, out remainingBytes, out invalid);
+            }
+            else
+            {
+                remainingBytes = SkipOpaqueType2(extensionData, out invalid);
+                return null;
+            }
+        }
+
+        private static string GetSniFromServerNameList(ReadOnlySpan<byte> serverNameListExtension, out ReadOnlySpan<byte> remainingBytes, out bool invalid)
+        {
+            // https://tools.ietf.org/html/rfc3546#section-3.1
+            // struct {
+            //     ServerName server_name_list<1..2^16-1>
+            // } ServerNameList;
+            // ServerNameList is an opaque type (length of sufficient size for max data length is prepended)
+            const int ServerNameListOffset = sizeof(ushort);
+
+            if (serverNameListExtension.Length < ServerNameListOffset)
+            {
+                remainingBytes = ReadOnlySpan<byte>.Empty;
+                invalid = true;
+                return null;
+            }
+
+            int serverNameListLength = BinaryPrimitives.ReadUInt16BigEndian(serverNameListExtension);
+            ReadOnlySpan<byte> serverNameList = serverNameListExtension.Slice(ServerNameListOffset);
+
+            if (serverNameListLength > serverNameList.Length)
+            {
+                remainingBytes = ReadOnlySpan<byte>.Empty;
+                invalid = true;
+                return null;
+            }
+
+            remainingBytes = serverNameList.Slice(serverNameListLength);
+            ReadOnlySpan<byte> serverName = serverNameList.Slice(0, serverNameListLength);
+
+            return GetSniFromServerName(serverName, out invalid);
+        }
+
+        private static string GetSniFromServerName(ReadOnlySpan<byte> serverName, out bool invalid)
+        {
+            // https://tools.ietf.org/html/rfc3546#section-3.1
+            // struct {
+            //     NameType name_type;
+            //     select (name_type) {
+            //         case host_name: HostName;
+            //     } name;
+            // } ServerName;
+            // ServerName is an opaque type (length of sufficient size for max data length is prepended)
+            const int ServerNameLengthOffset = 0;
+            const int NameTypeOffset = ServerNameLengthOffset + sizeof(ushort);
+            const int HostNameStructOffset = NameTypeOffset + sizeof(NameType);
+
+            if (serverName.Length < HostNameStructOffset)
+            {
+                invalid = true;
+                return null;
+            }
+
+            int hostNameStructLength = BinaryPrimitives.ReadUInt16BigEndian(serverName) - sizeof(NameType);
```

Comment:
Nit: add comment saying this could underflow, but it's ok due to the equality check below.